### PR TITLE
Extend currencies with subunit

### PR DESCRIPTION
**Pull request by frederikbosch**
Format currencies with subunits by using a subunit provider. Built-in providers are Constant and ISO. A follow-up PR might refactor `ISOCurrencies` and `ISOProvider` to use a single `ISORepository` or `ISOFactory`.

Questions:
1. Is `Provider` the correct name in this context?
2. What do you feel about the architectural decisions?
3. What about namespaces? Is everything in the correct place?

Please give me feedback on those important issues. Afterwards we can discuss secondary style issues. /cc @sagikazarmark 

Fixes: #221 #232 #233 #223.

### CODE REVIEW

**Comment by sagikazarmark** [comment]
I would rather create a repository interface extension: `CurrenciesWithSubunit`

``` php
function getSubunitFor(Currency $currency);
```

By definition a subunit should always be an integer value and I think we should fall back to 0 if we have no information.

**Comment by pamil** [comment]
Anyway, we should stick to either `subunit` or `subUnit` :) The first one looks better for me.

**Comment by frederikbosch** [comment]
@sagikazarmark Why do want to extend another interface?

**Comment by frederikbosch** [comment]
@pamil That can easily be renamed. I can do that.

**Comment by frederikbosch** [comment]
@sagikazarmark Actually, that is not completely different from what I am proposing. Just different names.

**Comment by sagikazarmark** [comment]
Well, it does not have to be an extension, but IMO we should implement these interfaces in monolithic repositories. Kind of breaks SRP, but it's repository pattern.

**Comment by sagikazarmark** [comment]
No, it's not different, but fits into the domain better IMHO.

**Comment by frederikbosch** [comment]
@sagikazarmark Refactoring... give me two minutes.

**Comment by frederikbosch** [comment]
@sagikazarmark Is this any better?

**Comment by frederikbosch** [comment]
@pamil What do you think?

**Comment by frederikbosch** [comment]
`CurrenciesWithSubunit` might be better of in `Money` namespace instead of the `Money\Currencies` namespace since `Currencies` is also in the `Money` namespace.

File: spec/Formatter/IntlMoneyFormatterSpec.php [hunk 1]
```diff
@@ -24,13 +25,13 @@ function it_is_a_money_formatter()
...
-    function it_formats_money(\NumberFormatter $numberFormatter)
+    function it_formats_money(\NumberFormatter $numberFormatter, CurrenciesWithSubunit $currenciesWithSubunit)
     {
-        $numberFormatter->getAttribute(\NumberFormatter::FRACTION_DIGITS)->willReturn(2);
-        $numberFormatter->formatCurrency('0.01', 'EUR')->willReturn('€1.00');
-
         $money = new Money(1, new Currency('EUR'));
 
+        $numberFormatter->formatCurrency('0.01', 'EUR')->willReturn('€1.00');
+        $currenciesWithSubunit->getSubunitsFor($money->getCurrency())->willReturn(2);
```
Comment (sagikazarmark) [inline spec/Formatter/IntlMoneyFormatterSpec.php]: Hm, I am not sure what is gramatically correct: subunit or subunitS. The first seems to be more logical to me. Can you check which is correct?

File: src/Currencies/ConstantSubunits.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * @var int
+     */
+    private $subUnits;
+
+    /**
+     * @param int $subUnits
+     */
+    public function __construct($subUnits)
```
Comment (sagikazarmark) [inline src/Currencies/ConstantSubunits.php]: As per the spec, shouldn't this default to 0?

File: src/Currencies/CurrenciesWithSubunit.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Money\Currencies;
+
+use Money\Currency;
+
+interface CurrenciesWithSubunit
```
Comment (sagikazarmark) [inline src/Currencies/CurrenciesWithSubunit.php]: author block and docblock

File: src/Currencies/ISOCurrencies.php [hunk 4]
```diff
@@ -31,6 +32,25 @@ public function contains(Currency $currency)
...
+     * {@inheritdoc}
+     */
+    public function getSubunitsFor(Currency $currency)
+    {
+        if (null === self::$currencies) {
+            self::$currencies = $this->loadCurrencies();
+        }
+
+        if (!isset(self::$currencies[$currency->getCode()])) {
+            throw new UnknownISOCurrencyException('Cannot find ISO currency '.$currency->getCode());
```
Comment (sagikazarmark) [inline src/Currencies/ISOCurrencies.php]: UnkownCurrencyException would be better? Could be used by other repositories too.

File: src/Currencies/ISOCurrencies.php [hunk 5]
```diff
@@ -31,6 +32,25 @@ public function contains(Currency $currency)
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getSubunitsFor(Currency $currency)
+    {
+        if (null === self::$currencies) {
+            self::$currencies = $this->loadCurrencies();
+        }
+
+        if (!isset(self::$currencies[$currency->getCode()])) {
```
Comment (sagikazarmark) [inline src/Currencies/ISOCurrencies.php]: Hm, should we do this or always return with zero if a currency is not found?

File: src/Formatter/IntlMoneyFormatter.php [hunk 6]
```diff
@@ -16,13 +17,19 @@
      * @var \NumberFormatter
      */
     private $formatter;
+    /**
+     * @var CurrenciesWithSubunit
+     */
+    private $currenciesWithSubunit;
```
Comment (sagikazarmark) [inline src/Formatter/IntlMoneyFormatter.php]: IMO $currencies is enough

**Comment by frederikbosch** [comment]
@sagikazarmark Since you are already giving detailed comments I assume you like this design better then the initial one?

**Comment by sagikazarmark** [comment]
Your assumption is correct. :wink:

**Comment by frederikbosch** [comment]
Haha, ok. I will fix those issues and think about the default of zero. Don't know for sure if that is what we want.

**Comment by frederikbosch** [comment]
Just updated the PR. I did not got for the default subunit of zero. Rather I added a `FallbackSubunit` class that can catch errors from a delegated `CurrenciesWithSubunit` a then use a fallback from another `CurrenciesWithSubunit`. Basically, the fallback will act as a decorator.

**Comment by frederikbosch** [comment]
Also went for `subunit`: singular.

**Comment by frederikbosch** [comment]
By the way, I have not created spec tests for the different `CurrenciesWithSubunit` because I want to have an agreement on the architecture. I will add those.

**Comment by frederikbosch** [comment]
Actually I think this solution it far better than others we discussed before. Thanks for the suggestion @sagikazarmark. I will leave this here for some days. Without objections this will be merged somewhere next week.

**Comment by frederikbosch** [comment]
My reasoning for not using a subunit default is as follows.

1. The `ISOCurrencies` class is only aware of the how and what of ISO currencies. A default is only appropriate if there is a ISO default. I do not believe there is one. Moreover, it would be weird when `contains` returns false and `getSubunitFor` actually returns a number. So it should indeed throw an exception.
2. `ConstantSubunit` is a class where you - if you use it - can inject the number zero without any problem. That is the whole idea of the class.

Hence, we need a third class: `FallbackSubunit` that applies a constant when the delegated is failing.
```php
<?php
use Money\Currencies\ConstantSubunit;
use Money\Currencies\FallbackSubunit;
use Money\Currencies\ISOCurrencies;
use Money\Formatter\IntlMoneyFormatter;

$currencies = new ISOCurrencies();

$formatter = new IntlMoneyFormatter(
  new \NumberFormatter(...),
  new FallbackSubunit(
    $currencies,
    new ConstantSubunit(0)
  )
);
```

**Comment by Sharom** [comment]
I think, that we can add `CurrencySpecification` value.

`CurrencySpecification` should contain information about subunits, name of subunit and so on.

Also we should provide `CurrencyRepositoryInterface`:

```php
interface CurrencySpecificationRepositoryInterface
{
    /**
     * @param Currency $currency
     * @return CurrencySpecification|null
     */
     public function findCurrencySpecification(Currency $currency);
}
```

After that we can inject an instance of `CurrencySpecificationRepositoryInterface` into the formatter/parser.

And `CurrencyRepository` can provide currency by currency specifications =)

**Comment by frederikbosch** [comment]
@Sharom I took your idea and merged into the branch. It is little different from your proposal but I definitely like the interface `CurrenciesSpecification` better than `CurrenciesWithSubunit`. The new value object `Specification` should throw exception I guess instead of returning null when it does not know a specific value (numeric code or name of a currency). But I'd like to hear first how @sagikazarmark and @pamil think about this compared to what we had before.

**Comment by frederikbosch** [comment]
Now, I think that CurrenciesSpecification should be moved into Currencies. So Currencies interface has two methods: contains and specify. If the repository contains the currency, it must be able to specify its characteristics. Right?

**Comment by Sharom** [comment]
> Now, I think that CurrenciesSpecification should be moved into Currencies. So Currencies interface has two methods: contains and specify. If the repository contains the currency, it must be able to specify its characteristics. Right?

@frederikbosch I am not sure. I think that it's better to have separated interfaces. `Currency` and `CurrencySpecification` are different things. Therefore a repository of currencies and repository of specifications are different too. Two responsibilities of a one interface smell bad.

But currency can be created by a specification. In this case `Currencies` can use repository of specifications.

I suggest not to rush to a final decision. Let's think about it.

File: src/CurrenciesSpecification.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+namespace Money;
+
+use Money\Currencies\Specification;
+
+/**
+ * Specifies information about a currency.
+ *
+ * @author Frederik Bosch <f.bosch@genkgo.nl>
+ */
+interface CurrenciesSpecification
```
Comment (Sharom) [inline src/CurrenciesSpecification.php]: I think, that it's better to rename this interface like  `CurrencySpecifications`, isn't it?

File: src/CurrenciesSpecification.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+{
+    /**
+     * @param Currency $currency
+     *
+     * @return Specification
```
Comment (Sharom) [inline src/CurrenciesSpecification.php]: And I would rename `Specification` -> `CurrecySpecification`. I think that it will be clearer.

File: src/CurrenciesSpecification.php (hunk 7 above)
Comment (Sharom) [inline src/CurrenciesSpecification.php]: I am not sure in choice of name for this interface.

**Comment by sagikazarmark** [comment]
> Now, I think that CurrenciesSpecification should be moved into Currencies. So Currencies interface has two methods: contains and specify. If the repository contains the currency, it must be able to specify its characteristics. Right?

This is Multiple Responsibility Principle too.

Can you give me some more usecases for a Specification object? Sounds like a generic metadata store, dunno if we need that.

**Comment by sagikazarmark** [comment]
I am going to turn of the automatic MR creation from StyleCI. It's annoying.

**Comment by frederikbosch** [comment]
@sagikazarmark Now I closed the other PR and got back here. And I also tried to work with `CurrenciesWithSubunit` like you are proposing. It simply does not work within our domain to have it as a new interface.

1. How should AggregateCurrencies look like when we have `CurrenciesWithSubunit`?
2. How should CachedCurrencies look like when we have `CurrenciesWithSubunit`?

While trying to make it work I found that the solutions can be.

1. We will end-up with dirty `instanceof` checks if we implement `CurrenciesWithSubunit` in those classes.
2. Or we have to create new classes `AggregateCurrenciesWithSubunit` and `CachedCurrenciesWithSubunit`. That becomes a nightmare.
3. We leave those classes untouched and we accept that there no aggregation and cache for subunit.

Basically decorating `Currencies` | `CurrenciesWithSubunit` becomes very hard if we add another interface for subunits. We should keep it simple. `subunitFor` is a requirement of our domain and belongs to something our repository should provide by default. Now if you have a specific currency without a subunit, you create your own `Currencies` implementation.

```php
namespace My;

use Money\Currencies;

class 8g7bv87b786cv8zuvb786bv7zv76v implement Currencies {

  public function contains(Currency $currency) {
     return $currency->getCode() === '8g7bv87b786cv8zuvb786bv7zv76v';
  }

  public function subunitFor(Currency $currency) {
    return 0;
  }
}
```

@sagikazarmark If you still have time... ;) Please provide some feedback here.

**Comment by Sharom** [comment]
**One other case, related to subunits**

Sometimes we need to accomplish some rounding of money:

1. Let we have some thing with purchase price equal to `10.20 USD`;
2. And we want to sell it with ~12% profit: `10.20 USD * 12% = 11.42 USD`;
3. But price `11.42 USD` looks ugly, `10.50 USD` looks better!

This functionality may be implemented on `Money` in the future.
And we need information about subunits to be able do this.

**Comment by frederikbosch** [comment]
I am going to wait until Travis finishes this one. I will squash it and merge it. There were many iterations and many experiments. I have done as much as I could with all the feedback. Now it is decision time. We have to move on.

**Comment by frederikbosch** [comment]
This PR includes formatting and parsing with rounding. That means parsing $1000.005 for a currency with subunit of 2 will become `Money::US(100001)` and formatting  `Money::US(5055)` with subunit of 2 and formatting fraction digits of 0 will become '$51'.

**Comment by frederikbosch** [comment]
This PR also includes `Currencies` extending `\IteratorAggregate`. So one we can iterate over every currency code inside a currency repository. So `Currencies` actually provides currency codes.

**Comment by frederikbosch** [comment]
This PR does not include a solution to a name of a currency. Iteration via `\IteratorAggregate` allows to see which currency codes are available. The library user can translate a currency code to a currency name via its own objects.

**Comment by frederikbosch** [comment]
Thanks @sagikazarmark @pamil @Sharom and others that helped giving the feedback.

**Comment by sagikazarmark** [comment]
I am not entirely happy to have `subunitFor` in the Currencies interface. Subunit is not a core part of Currencies. I am still on the side of a separate interface. I don't care if both are implemented in the same class, but they seem to be separate responsibilities.

> How should AggregateCurrencies look like when we have CurrenciesWithSubunit

The answer is simple: it shouldn't change. Since they are different responsibilities, you need to use two different objects: one Currencies instance decorated with AggregateCurrencies and an other one is the Subunit repo. (If it was DI, we would have two separate services).

On the other hand, it would probably increase complexity which might not worth the effort.

A few cosmetic points:

- methods names are better when they are in imperative: `getSubunitFor`. Contains is kind of an exception
- document better when UnkownCurrencyException should be thrown (when repo does not contain currency via contains?)

**Comment by frederikbosch** [comment]
@sagikazarmark Thanks for the feedback. I understand it very well. I will have a look at the `UnknownCurrencyException`.
